examples/contrib/timeseries/hmc_gp.py

```diff
@@ -0,0 +1,65 @@
+import argparse
+
+import numpy as np
+import torch
+
+import pyro
+import pyro.distributions as dist
+from pyro.infer.mcmc import NUTS
+from pyro.infer.mcmc.api import MCMC
+from pyro.contrib.timeseries import IndependentMaternGP
+
+from gp_models import download_data
+
+
+pyro.enable_validation(__debug__)
+
+
+def main(args):
+    download_data()
+    data = np.loadtxt('eeg.dat', delimiter=',', skiprows=19)
+    print("[raw data shape] {}".format(data.shape))
+    data = torch.tensor(data[::20, :-1]).double()
+    data = data[:100, :1]  # choose a single output dimension
+    print("[data shape after thinning] {}".format(data.shape))
+
+    T, obs_dim = data.shape
+
+    # standardize data
+    data_mean = data.mean(0)
+    data -= data_mean
+    data_std = data.std(0)
+    data /= data_std
+
+    torch.manual_seed(args.seed)
+
+    gp = IndependentMaternGP(nu=1.5, obs_dim=obs_dim).double()
+
+    def model(*args):
+        del gp.kernel.log_length_scale
+        del gp.kernel.log_kernel_scale
+        del gp.log_obs_noise_scale
+        prior_dist = dist.Normal(0.0, 2 * torch.ones(obs_dim)).to_event(1)
+        setattr(gp.kernel, "log_length_scale",
+                pyro.sample("log_length_scale", prior_dist))
+        setattr(gp.kernel, "log_kernel_scale",
+                pyro.sample("log_kernel_scale", prior_dist))
+        setattr(gp, "log_obs_noise_scale",
+                pyro.sample("log_obs_noise_scale", prior_dist))
```

Comment:
@fritzo i would to refactor contrib.timeseries so that this is less gross. as previously discussed this may need something like `LiftedParameter`

Reply:
Very interesting pattern, I'll try to address this in #2134.

Reply:
After #2134 merges, you could try this:
1. Modify pyro.contrib.timeseries.gp:
```diff
diff --git a/pyro/contrib/timeseries/gp.py b/pyro/contrib/timeseries/gp.py
index 5deb8c0f..e53ca9af 100644
--- a/pyro/contrib/timeseries/gp.py
+++ b/pyro/contrib/timeseries/gp.py
@@ -27,6 +27,8 @@ class IndependentMaternGP(TimeSeriesModel):
     :param torch.Tensor log_obs_noise_scale_init: optional initial values for the observation noise scale
         given as a ``obs_dim``-dimensional tensor
     """
+    Kernel = MaternKernel
+
     def __init__(self, nu=1.5, dt=1.0, obs_dim=1,
                  log_length_scale_init=None, log_kernel_scale_init=None,
                  log_obs_noise_scale_init=None):
@@ -40,9 +42,9 @@ class IndependentMaternGP(TimeSeriesModel):

         super().__init__()

-        self.kernel = MaternKernel(nu=nu, num_gps=obs_dim,
-                                   log_length_scale_init=log_length_scale_init,
-                                   log_kernel_scale_init=log_kernel_scale_init)
+        self.kernel = self.Kernel(nu=nu, num_gps=obs_dim,
+                                  log_length_scale_init=log_length_scale_init,
+                                  log_kernel_scale_init=log_kernel_scale_init)

         self.log_obs_noise_scale = nn.Parameter(log_obs_noise_scale_init)
```

2. Use `PyroModule` and `PyroSample` here:
```py
class RandomMaternKernel(MaternKernel, PyroModule):
    pass

class RandomGP(IndependentMaternGP, PyroModule):
    Kernel = RandomMaternKernel
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        prior_dist = dist.Normal(0.0, 2 * torch.ones(obs_dim)).to_event(1)
        self.kernel.log_length_scale = PyroSample(prior_dist)
        self.kernel.log_kernel_scale = PyroSample(prior_dist)
        self.kernel.log_obs_noise_scale = PyroSample(prior_dist)
```

Reply:
tried to do this but i must be missing something because i'm getting
```
  File "/home/jankowiak/miniconda3/envs/torch/lib/python3.7/site-packages/torch/nn/modules/module.py", line 604, in __setattr__
    .format(torch.typename(value), name))
TypeError: cannot assign 'pyro.nn.module.PyroSample' as parameter 'log_length_scale' (torch.nn.Parameter or None expected)
```

Reply:
Looks like you modified `LinearlyCoupledMaternGP` instead of `IndependentMaternGP`. First try this:
```diff
diff --git a/pyro/contrib/timeseries/gp.py b/pyro/contrib/timeseries/gp.py
index 034034d2..11a11dad 100644
--- a/pyro/contrib/timeseries/gp.py
+++ b/pyro/contrib/timeseries/gp.py
@@ -27,6 +27,8 @@ class IndependentMaternGP(TimeSeriesModel):
     :param torch.Tensor log_obs_noise_scale_init: optional initial values for the observation noise scale
         given as a ``obs_dim``-dimensional tensor
     """
+    Kernel = MaternKernel
+
     def __init__(self, nu=1.5, dt=1.0, obs_dim=1,
                  log_length_scale_init=None, log_kernel_scale_init=None,
                  log_obs_noise_scale_init=None):
@@ -40,9 +42,9 @@ class IndependentMaternGP(TimeSeriesModel):

         super().__init__()

-        self.kernel = MaternKernel(nu=nu, num_gps=obs_dim,
-                                   log_length_scale_init=log_length_scale_init,
-                                   log_kernel_scale_init=log_kernel_scale_init)
+        self.kernel = self.Kernel(nu=nu, num_gps=obs_dim,
+                                  log_length_scale_init=log_length_scale_init,
+                                  log_kernel_scale_init=log_kernel_scale_init)

         self.log_obs_noise_scale = nn.Parameter(log_obs_noise_scale_init)
```
To resolve remaining errors, define a `.forward()` method and ensure your kernels are accessed through that method.

Reply:
@fritzo thanks this now works. pretty neat 😃